### PR TITLE
Ports Vendor Icon Display instead of Color from Yogs, with some fixes.

### PR DESCRIPTION
**Pull request by KomradeSpectre**
## About The Pull Request
Port of https://github.com/yogstation13/Yogstation-TG/pull/2549
which is a port of https://github.com/Citadel-Station-13/Citadel-Station-13/pull/7442
which is a port of https://github.com/OracleStation/OracleStation/pull/50 (jesus)
All of which contain improvements on the previous tier of the port, including mine as well.

This port visualizes items in all machines handled by _vending.dm, allowing a person to see what they're buying before buying it. This is very useful for clothing vendors for example, and much better than the previous system that had vending machines choose from three colors randomly.

Here's a preview of it in action. (Edit: New Picture with all the fixes)
image

Credits to:

- AndrewMontagne (Oracle, original implementation)
- FlattestGuitar (Citadel)
- nichlas0010 (Yogstation)

## Why It's Good For The Game

Adds the ability to preview what the item you're buying looks like before purchasing it. This allows for a more informed decision when purchasing clothes and other items, instead of wasting stock and throwing it away somewhere where other crew members wouldn't be able to use it unless they go find it (though you can still do this if you want to be a dick, I guess.)

Besides, it looks nice.

## Changelog
:cl: AndrewMontague, FlattestGuitar, nichlas0010, TheMythicGhost
tweak: Changes vending machine selections to now display their contents icon instead of just their name and a random color.
/:cl:

### CODE REVIEW

**Comment by Tlaltecuhtli** [comment]
can this be generalized to fridges?

**Comment by KomradeSpectre** [comment]
Smartfridges use TGUI, this uses the browser interface (HTML). If anyone would care to give any clue how TGUI even functions, please feel free.

File: code/modules/vending/_vending.dm [hunk 1]
```diff
@@ -179,7 +188,7 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 		if(!start_empty)
 			R.amount = amount
 		R.max_amount = amount
-		R.display_color = pick("#ff8080","#80ff80","#8080ff")
+		//R.display_color = pick("#ff8080","#80ff80","#8080ff") - We're using icon instead of color, but this is here incase someone wants to easily revert the change.
```
Comment (ShizCalev) [inline code/modules/vending/_vending.dm]: just delete this, we can see it in git history.

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Tlaltecuhtli** [comment]
>html looks better than tgui

File: code/modules/vending/_vending.dm [hunk 2]
```diff
@@ -347,15 +355,14 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
...
 			if(R.amount > 0 && ((C && C.registered_account && onstation) || (!onstation && isliving(user))))
-				dat += "<a href='byond://?src=[REF(src)];vend=[REF(R)]'>Vend</a> "
+				dat += "<td align='right'><b>[R.amount]&nbsp;</b></td><td><a href='byond://?src=[REF(src)];vend=[REF(R)]'>Vend</a></td>"
 			else
-				dat += "<span class='linkOff'>Not Available</span> "
-			dat += "<font color = '[R.display_color]'><b>[sanitize(R.name)] ([price_listed])</b>:</font>"
-			dat += " <b>[R.amount]</b>"
-			dat += "</li>"
-		dat += "</ul>"
+				dat += "<td><span class='linkOff'>Not&nbsp;Available</span></td>"
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/vending/_vending.dm]: right align this so it doesn't have that weird shift.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/vending/_vending.dm]: ping me after so I can merge

Comment (KomradeSpectre) [reply code/modules/vending/_vending.dm]: Will quickly test to see if that actually aligns it to the right after the amount is gone, if so, I'll do a quick little edit and ping you.

**Comment by KomradeSpectre** [comment]
@ExcessiveUseOfCobblestone All fixed.
image

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
The 'revert'-related comments are unnecessary, we have git for that.

Other than that, I can think of some ways this might be optimized but if it works it works.

**Comment by KomradeSpectre** [comment]
If you know ways this can be optimized, shoot. Also, do you mean the comment indicating the start of icon instead of color and the end? I can remove those if you mean that.

**Comment by SpaceManiac** [comment]
The start and end comments are not needed, and just removing `var/display_color` rather than commenting it is fine.

We have a spritesheet system used by the R&D Console's icon display, so you could re-use that (see `code/modules/client/asset_cache.dm`), I think I would need to do some measuring to know whether it works better or worse than what you already have though.

**Comment by KomradeSpectre** [comment]
Righto, I'll remove the comments. I've already removed the commented out line (as indicated in one of my commits), but I'll get those comments out too (I only added them in as they were in on the origin port.)

**Comment by KomradeSpectre** [comment]
@SpaceManiac Done. Should be good to go?

**Review by SpaceManiac — APPROVED** [review]

**Pull request merged**